### PR TITLE
fix: optimize queries

### PR DESCRIPTION
**Pull request by chris13524**
# Description

Seems the `dead_letters_check` query takes a long time to run. And even when it's not spiking like this, it still takes the longest out of the other queries. Attempting to optimize it by not doing maths in the query, and instead leveraging the existing indexes.

Also removed an redundant `dead_letter_give_up_check` query for checking `created_at` value.

image

## How Has This Been Tested?

Automated tests

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

File: src/services/publisher_service/helpers.rs [hunk 1]
```diff
@@ -276,26 +279,10 @@ pub async fn dead_letters_check(
...
-        .bind(threshold.as_secs() as i64)
-        .bind(notification)
-        .fetch_one(postgres)
-        .await?;
-    if let Some(metrics) = metrics {
-        metrics.postgres_query("dead_letter_give_up_check", start);
-    }
-    Ok(row.0)
+) -> bool {
+    notification_created_at + threshold < Utc::now()
```
Comment (chris13524) [inline src/services/publisher_service/helpers.rs:287]: This ordering is easier to understand to me than `Utc::now() - notification_created_at > threshold`

Order the variables in a timeline, oldest to newest.

**Review by boidushya — APPROVED** [review]

**Review by xav — APPROVED** [review]

File: src/services/publisher_service/helpers.rs [hunk 2]
```diff
@@ -276,26 +279,10 @@ pub async fn dead_letters_check(
 }
 
 /// Checks for message is created more than threshold
-#[instrument(skip(postgres, metrics))]
-pub async fn dead_letter_give_up_check(
-    notification: Uuid,
+#[instrument]
+pub fn dead_letter_give_up_check(
```
Comment (geekbrother) [inline src/services/publisher_service/helpers.rs:283]: Good catch to refactor it without a second query!

**Review by geekbrother — APPROVED** [review]

**Comment by geekbrother** [comment]
I don't think this 14s spike in latency on the graph is because of the simple `SELECT` query in the dead letter check as we are selecting by the indexed field with a simple query.

**Comment by chris13524** [comment]
> I don't think this 14s spike in latency on the graph is because of the simple SELECT query in the dead letter check as we are selecting by the indexed field with a simple query.

Yeah not sure why this would cause it, it could be the result of other things. But it still seems to be poor-performing to some extent, and we can optimize it.

**Pull request merged**